tests/containers/validate_btrfs.pm

```
@@ -0,0 +1,79 @@
+use testapi;
+use containers::runtime;
+
+sub _sanity_test_btrfs {
+    my ($rt, $dev_path) = @_;
+    my $dockerfile_path = '/root/sle_base_image/docker_build';
+    my $btrfs_head      = '/tmp/subvolumes_saved';
+    $rt->info(property => 'Driver', value => 'btrfs');
+    $rt->build($dockerfile_path, 'huge_image');
+    assert_script_run "btrfs fi df $dev_path/btrfs/";
+    assert_script_run "ls -td $dev_path/btrfs/subvolumes/* | head -n 1 > $btrfs_head";
+    validate_script_output "df -h |grep var", sub { m/\/dev\/vda.+[1-6]\d?%/ };
+    $rt->up('huge_image', cmd => 'fallocate -l 8.7G bigfile.stty');
+    # partition should be full
+    validate_script_output "df -h |grep var",       sub { m/\/dev\/vda.+\s+(9[7-9]|100)%/ };
+    validate_script_output "btrfs fi df $dev_path", sub { m/^Data.+total=8.*GiB, used=8.*GiB/ };
+}
+
+sub _test_btrfs_balancing {
+    my ($dev_path) = shift;
+    assert_script_run qq(btrfs balance start $dev_path), timeout => 600;
+    assert_script_run "btrfs fi show $dev_path/btrfs";
+    validate_script_output "btrfs fi show $dev_path/btrfs", sub { m/devid\s+2.+20.00G.+1.\d+G.+\/dev\/vdb/ };
+}
+
+sub _test_btrfs_thin_partitioning {
+    my ($rt, $dev_path) = @_;
+    my $dockerfile_path = '/root/sle_base_image/docker_build';
+    my $btrfs_head      = '/tmp/subvolumes_saved';
+    $rt->build($dockerfile_path, 'thin_image');
+    # validate that new subvolume has been created. This should be improved.
+    assert_script_run qq{test \$(ls -td $dev_path/btrfs/subvolumes/* | head -n 1) == \$(cat $btrfs_head)};
+    validate_script_output "btrfs fi df $dev_path", sub { m/^Data.+total=8.*GiB, used=8.*GiB/ };
+}
+
+sub _test_btrfs_device_mgmt {
+    my ($rt, $dev_path) = @_;
+    my $tw         = 'registry.suse.de/home/favogt/jumpmeta/images_tw/opensuse/tumbleweed:latest';
+    my $btrfs_head = '/tmp/subvolumes_saved';
+    # Due to disk space this should fail
+    assert_script_run qq(docker pull $tw || echo pull_failed);
```

What does this do? Do you expect a failure? Is it a failure if it doesn't fail?
Also:
```suggestion
    $rt->_rt_assert_script_run('pull $tw || echo pull_failed');
```

---

Yes, failure is expected. Simulates disk fullness. so the `pull` should fail and ensure that then it works when we add a new device.

---

i think the use of `_rt_assert_script_run` breaks the encapsulation. it is used as private method in the class. This ideally should be handled inside in the `pull`. Do you think i need to cover this? i am not sure how often will call pull expected a failure, and that why i did not include it

---

What do you think about

    assert_script_run("! docker pull $tw");

IMHO grepping for `pull_failed` is not really specific and then it makes more sense to check for the return code than this output.
Only a suggestion though, I think it also will work fine as it is :-)

---

If `docker pull $tw` doesn't fail, the test should fail or?

---

Regarding direct use of `_rt_assert_script_run()`, if you treat this private, then ok. ( I have actually something like https://github.com/cfconrad/os-autoinst-distri-opensuse/commit/df70d67668568320ef67a67c2da9a66450f32cfb in my mind. But I would just comeup with a other PR)

Beside this, I think it doesn't break anything. And I would just allow to call it.

---

> IMHO grepping for `pull_failed` is not really specific and then it makes more sense to check for the return code than this output.
> Only a suggestion though, I think it also will work fine as it is :-)

What I'm mistaken, I see a `echo` no `grep` and with this, we just hide the error case

---

let me try with the negation. FYI failing in this step it should not make the test die. it is just a validation that it cannot actually pull anything.

---

True there was no `grep`, it must be eye memory or something ;-)

---

`grep` is not working. because of `pipefail`

---

addressed on 8d47f4130 using negation. Thanks guys for suggestion